File: P6_Box_plot/program.cpp
#include<bits/stdc++.h>
using namespace std;

vector<double> getData(){
    vector<double> result;

    ifstream inpfile("./input.csv");
    string line;

    getline(inpfile,line);

    while(getline(inpfile,line)){
        double val;
        val=stod(line);
        result.push_back(val);
    }

    return result;
};

double getMedian(int n,vector<double>& data){

    // for(auto it:data){
    //     cout<<it<<" ";
    // }
    // cout<<endl;
    // cout<<n<<endl;
    // cout<<data[n/2]<<endl;
    // cout<<data[n/2-1]<<endl;
    // cout<<data[n/2+1]<<endl;
    return n%2?(double)(data[n/2-1]):(((double)(data[n/2-1])+(data[n/2]))/2.0);
}

void boxPlot(vector<double>& data){
    sort(data.begin(),data.end());

    int n=data.size();

    double min_ele=data[0];
    double max_ele=data[n-1];

    double median=getMedian(n,data);

    int q1_index=int(n*0.25);
    int q3_index=int(n*0.75);

    double Q1=data[q1_index];
    double Q3=data[q3_index];

    double IQR=Q3-Q1;

    double lower_bound=Q1-1.5*IQR;
    double upper_bound=Q3+1.5*IQR;

    auto outliers=[&](){
        vector<double> outliers;

        for(auto it:data){
            if(it<lower_bound || it>upper_bound){
                outliers.push_back(it);
            }
        }  

        return outliers;
    };

    vector<double> outliersData=outliers();


    cout<<"Lower Bound: "<<lower_bound<<endl;
    cout<<"Q1: "<<Q1<<endl;
    cout<<"Median: "<<median<<endl;
    cout<<"Q3: "<<Q3<<endl;
    cout<<"Upper Bound: "<<upper_bound<<endl;
    cout<<"IQR: "<<IQR<<endl;
    cout<<"Outliers: "<<endl;

    for(auto it:outliersData){
        cout<<it<<" ";
    }
    cout<<endl;
}

int main(){
    vector<double> data=getData();

   boxPlot(data);
}

File: P6_Box_plot/program.py
import pandas as pd
import statistics

def five_number_summary(data):
    clean_data = [x for x in data if pd.notna(x)]
    if not clean_data:
        return None
    sorted_data = sorted(clean_data)
    n = len(sorted_data)
    mini = sorted_data[0]
    maxi = sorted_data[-1]
    median = statistics.median(sorted_data)
    q1_index = int(n * 0.25)
    q3_index = int(n * 0.75)
    Q1 = sorted_data[q1_index]
    Q3 = sorted_data[q3_index]
    IQR = Q3 - Q1
    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR
    outliers = [x for x in sorted_data if x < lower_bound or x > upper_bound]
    return mini, Q1, median, Q3, maxi, IQR, lower_bound, upper_bound, outliers

def main():
    file_path = '../Student_Dropout.csv'
    try:
        data = pd.read_csv(file_path)
        
        numeric_columns = data.select_dtypes(include=['int64', 'float64']).columns
        print("\navailable numeric columns for analysis:")
        for i, col in enumerate(numeric_columns, 1):
            print(f"{i}. {col}")
        
        while True:
            try:
                choice = int(input("\nenter the index of the column: "))
                if 1 <= choice <= len(numeric_columns):
                    selected_column = numeric_columns[choice-1]
                    break
                else:
                    print("invalid column")
            except ValueError:
                print("enter valid column")
        
        data[selected_column] = pd.to_numeric(data[selected_column], errors='coerce')
        values = data[selected_column].dropna().tolist()
        
        if not values:
            print(f"\ninvalid data: {selected_column}")
        else:
            result = five_number_summary(values)
            if result:
                mini, Q1, median, Q3, maxi, IQR, low, high, outliers = result
                print(f"\nfive number summary for {selected_column}:")
                print("minimum:", mini)
                print("Q1:", Q1)
                print("median:", median)
                print("Q3:", Q3)
                print("maximum:", maxi)
                print("IQR:", IQR)
                print("number of outliers:", len(outliers))
                print("lower bound for outliers:", low)
                print("upper bound for outliers:", high)
                
                if outliers and len(outliers) <= 10:
                    print("Outliers:", outliers)
            else:
                print(f"error occured {selected_column}")
    
    except FileNotFoundError:
        print(f"error occured {file_path}")
    except Exception as e:
        print(f"error occurred: {str(e)}")

if __name__ == "__main__":
    main()